File: quant/correlation.py
#!/usr/bin/env python3
"""
Rolling correlation + crisis-correlation check.

Tests the hypothesis "this asset diversifies my portfolio" —
calculates correlation in normal vs. stressed periods.
A portfolio that looks diversified in calm markets often compresses to ~1 in crashes.

Usage:
    python3 correlation.py --returns-csv returns.csv --window 30
    python3 correlation.py --returns-csv crypto_daily.csv --window 30 --crisis-threshold -0.05
"""
import argparse
try:
    from ._validation import number, series, integer
except ImportError:
    from pathlib import Path
    import sys
    sys.path.insert(0, str(Path(__file__).resolve().parent))
    from _validation import number, series, integer

import csv
import json
import math
import sys


def corr(x, y):
    series(x, "x", minimum_length=0)
    series(y, "y", minimum_length=0)
    if len(x) != len(y):
        raise ValueError("paired series must have equal lengths")
    n = len(x)
    if n < 2 or all(v == x[0] for v in x) or all(v == y[0] for v in y):
        return 0.0
    mx = sum(x) / n
    my = sum(y) / n
    num = sum((x[i] - mx) * (y[i] - my) for i in range(n))
    dx = math.sqrt(sum((v - mx) ** 2 for v in x))
    dy = math.sqrt(sum((v - my) ** 2 for v in y))
    if dx == 0 or dy == 0:
        raise ValueError("nonconstant variance is below floating-point resolution")
    return number(num / (dx * dy), "correlation")


def correlation_value(x, y):
    """JSON report uses null for unidentified Pearson correlations instead of a zero sentinel."""
    value = corr(x, y)
    if len(x) < 2 or all(v == x[0] for v in x) or all(v == y[0] for v in y):
        return None
    return round(value, 4)


def parse_returns_csv(lines, *, minimum_rows=2, minimum_assets=2):
    """Accept one all-text header, then finite rectangular numeric rows; never drop bad data."""
    rows = []
    first = True
    for line_number, row in enumerate(csv.reader(lines), 1):
        if not row:
            continue
        parsed = []
        for value in row:
            try:
                parsed.append(float(value))
            except ValueError:
                parsed.append(None)
        if first and all(value is None for value in parsed):
            first = False
            continue
        first = False
        if any(value is None for value in parsed):
            raise ValueError(f"malformed numeric CSV row {line_number}")
        series(parsed, f"CSV row {line_number}")
        if rows and len(parsed) != len(rows[0]):
            raise ValueError(f"ragged CSV row {line_number}")
        rows.append(parsed)
    if len(rows) < minimum_rows or (rows and len(rows[0]) < minimum_assets):
        raise ValueError(f"need at least {minimum_rows} observations and {minimum_assets} asset columns")
    return rows


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--returns-csv", required=True)
    ap.add_argument("--asset-names", help="comma-separated")
    ap.add_argument("--window", type=int, default=30)
    ap.add_argument("--crisis-threshold", type=float, default=-0.05,
                    help="day is 'crisis' if benchmark (col 0) return < threshold")
    args = ap.parse_args()

    integer(args.window, "window", minimum=2)
    number(args.crisis_threshold, "crisis threshold")
    with open(args.returns_csv) as f:
        cols = parse_returns_csv(f)

    n_rows = len(cols)
    n_assets = len(cols[0])
    names = args.asset_names.split(",") if args.asset_names else [f"a{i}" for i in range(n_assets)]
    if len(names) != n_assets or any(not name.strip() for name in names) or len(set(names)) != len(names):
        raise ValueError("asset names must be nonempty, unique and match the columns")
    series = [[row[j] for row in cols] for j in range(n_assets)]

    # full-sample pairwise correlation matrix
    full_corr = {}
    for i in range(n_assets):
        for j in range(i + 1, n_assets):
            full_corr[f"{names[i]}__{names[j]}"] = correlation_value(series[i], series[j])

    # crisis-only correlation: days where benchmark (asset 0) < threshold
    crisis_idx = [k for k in range(n_rows) if series[0][k] < args.crisis_threshold]
    crisis_corr = {}
    if len(crisis_idx) > 5:
        for i in range(n_assets):
            for j in range(i + 1, n_assets):
                crisis_corr[f"{names[i]}__{names[j]}"] = correlation_value(
                    [series[i][k] for k in crisis_idx], [series[j][k] for k in crisis_idx]
                )

    # correlation compression flag: crisis - normal
    compression = {}
    for k in full_corr:
        if full_corr[k] is not None and crisis_corr.get(k) is not None:
            compression[k] = round(crisis_corr[k] - full_corr[k], 4)

    # rolling window correlation vs. benchmark (last window)
    rolling = {}
    if n_rows >= args.window:
        for i in range(1, n_assets):
            tail_bench = series[0][-args.window:]
            tail_asset = series[i][-args.window:]
            rolling[f"{names[i]}_vs_{names[0]}"] = correlation_value(tail_bench, tail_asset)

    print(json.dumps({
        "n_observations": n_rows,
        "n_assets": n_assets,
        "assets": names,
        "window": args.window,
        "crisis_threshold": args.crisis_threshold,
        "n_crisis_days": len(crisis_idx),
        "undefined_correlation_convention": "null means insufficient observations or zero variance; excluded from compression",
        "full_sample_correlation": full_corr,
        "crisis_correlation": crisis_corr,
        "correlation_compression": compression,
        "rolling_correlation_last_window": rolling,
        "interpretation_hint": "Compression > 0.2 is a descriptive review flag, not a calibrated threshold or proof that diversification fails.",
    }, indent=2, allow_nan=False))


if __name__ == "__main__":
    main()

File: quant-jvm/src/main/kotlin/org/maxmoran/quant/Correlation.kt
package org.maxmoran.quant

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import java.io.File
import kotlin.math.sqrt
import kotlin.system.exitProcess

private val PRETTY_JSON = Json { prettyPrint = true }

/**
 * Deterministic port of `quant/correlation.py` — full-sample, crisis-only, and rolling
 * pairwise Pearson correlation with a compression flag.
 *
 * parity-contract.md §1.4 lists correlation under numerical linear algebra (1e-6), but the
 * reference performs no matrix decomposition — only pairwise sums — so this port meets the
 * stronger §1.1 deterministic regime (1e-10 raw doubles, exact rounded JSON).
 */

/** Pearson correlation; returns 0.0 for short series or zero variance, matching `corr`. */
fun corr(x: List<Double>, y: List<Double>): Double {
    require(x.size == y.size && x.all { it.isFinite() } && y.all { it.isFinite() }) { "paired series must be finite and have equal lengths" }
    val n = x.size
    if (n < 2 || x.all { it == x[0] } || y.all { it == y[0] }) return 0.0
    val mx = x.sum() / n
    val my = y.sum() / n
    var num = 0.0
    for (i in 0 until n) num += (x[i] - mx) * (y[i] - my)
    val dx = sqrt(x.sumOf { value ->
        val deviation = value - mx
        deviation * deviation
    })
    val dy = sqrt(y.sumOf { value ->
        val deviation = value - my
        deviation * deviation
    })
    require(dx != 0.0 && dy != 0.0) { "nonconstant variance is below floating-point resolution" }
    return (num / (dx * dy)).also { require(it.isFinite()) { "correlation overflow" } }
}

private fun correlationValue(x: List<Double>, y: List<Double>): Double? {
    val value = corr(x, y)
    if (x.size < 2 || x.all { it == x[0] } || y.all { it == y[0] }) return null
    return round4(value)
}

/** Accept one all-text header, then rectangular finite numeric rows; reject malformed data.
 * The line-oriented parser supports unquoted fields and quoted single-line fields without embedded commas.
 */
fun parseReturnsCsv(text: String): List<List<Double>> {
    val rows = mutableListOf<List<Double>>()
    var first = true
    for ((index, line) in text.lineSequence().withIndex()) {
        if (line.isEmpty()) continue
        val fields = line.split(",").map { it.trim().removeSurrounding("\"") }
        val values = fields.map { it.toDoubleOrNull() }
        if (first && values.all { it == null }) {
            first = false
            continue
        }
        first = false
        require(values.all { it != null && it.isFinite() }) { "malformed numeric CSV row ${index + 1}" }
        val numeric = values.filterNotNull()
        require(rows.isEmpty() || numeric.size == rows[0].size) { "ragged CSV row ${index + 1}" }
        rows.add(numeric)
    }
    return rows
}

/** Public rounded JSON value contract, field-for-field with `quant/correlation.py`. */
fun correlationOutput(
    cols: List<List<Double>>,
    assetNames: List<String>?,
    window: Int = 30,
    crisisThreshold: Double = -0.05,
): JsonObject {
    require(cols.size >= 2 && cols[0].size >= 2 && cols.all { row -> row.size == cols[0].size && row.all { it.isFinite() } }) { "need finite rectangular data with two observations and two assets" }
    require(window >= 2 && crisisThreshold.isFinite()) { "window must be >= 2 and crisis threshold finite" }
    val nRows = cols.size
    val nAssets = cols[0].size
    val names = assetNames ?: List(nAssets) { "a$it" }
    require(names.size == nAssets && names.all { it.isNotBlank() } && names.toSet().size == names.size) { "asset names must be nonempty, unique and match columns" }
    val series = List(nAssets) { j -> cols.map { it[j] } }

    val fullCorr = LinkedHashMap<String, Double?>()
    for (i in 0 until nAssets) {
        for (j in i + 1 until nAssets) {
            fullCorr["${names[i]}__${names[j]}"] = correlationValue(series[i], series[j])
        }
    }

    val crisisIdx = (0 until nRows).filter { series[0][it] < crisisThreshold }
    val crisisCorr = LinkedHashMap<String, Double?>()
    if (crisisIdx.size > 5) {
        for (i in 0 until nAssets) {
            for (j in i + 1 until nAssets) {
                crisisCorr["${names[i]}__${names[j]}"] = correlationValue(crisisIdx.map { series[i][it] }, crisisIdx.map { series[j][it] })
            }
        }
    }

    // The reference subtracts the already-rounded values, then rounds the difference again.
    val compression = LinkedHashMap<String, Double>()
    for ((key, full) in fullCorr) {
        if (full == null) continue
        val crisis = crisisCorr[key] ?: continue
        compression[key] = round4(crisis - full)
    }

    val rolling = LinkedHashMap<String, Double?>()
    if (nRows >= window) {
        for (i in 1 until nAssets) {
            rolling["${names[i]}_vs_${names[0]}"] =
                correlationValue(series[0].takeLast(window), series[i].takeLast(window))
        }
    }

    return buildJsonObject {
        put("n_observations", JsonPrimitive(nRows))
        put("n_assets", JsonPrimitive(nAssets))
        put("assets", buildJsonArray { names.forEach { add(JsonPrimitive(it)) } })
        put("window", JsonPrimitive(window))
        put("crisis_threshold", JsonPrimitive(crisisThreshold))
        put("n_crisis_days", JsonPrimitive(crisisIdx.size))
        put("undefined_correlation_convention", JsonPrimitive("null means insufficient observations or zero variance; excluded from compression"))
        put("full_sample_correlation", mapToJson(fullCorr))
        put("crisis_correlation", mapToJson(crisisCorr))
        put("correlation_compression", mapToJson(compression))
        put("rolling_correlation_last_window", mapToJson(rolling))
        put(
            "interpretation_hint",
            JsonPrimitive(
                "Compression > 0.2 is a descriptive review flag, not a calibrated threshold or proof that diversification fails."
            ),
        )
    }
}

private fun mapToJson(values: Map<String, Double?>): JsonObject = buildJsonObject {
    values.forEach { (key, value) -> put(key, value?.let { JsonPrimitive(it) } ?: JsonNull) }
}

/** Testable result of evaluating the correlation CLI contract. */
data class CorrelationCliResult(val exitCode: Int, val output: String)

/** Parse input and produce CLI output without terminating the process. */
fun evaluateCorrelation(args: Array<String>): CorrelationCliResult {
    return try {
        var csvPath: String? = null
        var assetNames: String? = null
        var window = 30
        var crisisThreshold = -0.05
        var index = 0

        fun optionValue(option: String): String {
            if (index + 1 >= args.size || args[index + 1].startsWith("--")) {
                throw IllegalArgumentException("$option requires a value")
            }
            index += 1
            return args[index]
        }

        while (index < args.size) {
            when (val arg = args[index]) {
                "--returns-csv" -> csvPath = optionValue(arg)
                "--asset-names" -> assetNames = optionValue(arg)
                "--window" -> window = optionValue(arg).toIntOrNull()
                    ?: throw IllegalArgumentException("--window must be an integer")
                "--crisis-threshold" -> crisisThreshold = optionValue(arg).toDoubleOrNull()
                    ?: throw IllegalArgumentException("--crisis-threshold must be a number")
                else -> throw IllegalArgumentException("unknown argument: $arg")
            }
            index += 1
        }

        val path = csvPath ?: throw IllegalArgumentException("need --returns-csv")
        // Python's "".split(",") yields [""], so an empty names flag still produces one name.
        val names = assetNames?.split(",")
        val output = correlationOutput(parseReturnsCsv(File(path).readText()), names, window, crisisThreshold)
        CorrelationCliResult(
            exitCode = 0,
            output = PRETTY_JSON.encodeToString(JsonObject.serializer(), output),
        )
    } catch (exception: Exception) {
        val message = exception.message?.lineSequence()?.firstOrNull()?.take(240) ?: "invalid input"
        CorrelationCliResult(
            exitCode = 1,
            output = buildJsonObject { put("error", JsonPrimitive(message)) }.toString(),
        )
    }
}

/** Module CLI runner — CSV in / JSON out, matching the Python entrypoint. */
fun runCorrelation(args: Array<String>) {
    val result = evaluateCorrelation(args)
    println(result.output)
    if (result.exitCode != 0) exitProcess(result.exitCode)
}
